File: src/composite/composite.rb
class Task # Component
  attr_reader :name

  def initialize(name)
    @name = name
  end

  def get_time_required
    0.0
  end
end

class CompositeTask < Task

  def initialize(name)
    super(name)
    @sub_tasks = []
  end

  def add_sub_task(task)
    @sub_tasks << task
  end

  def remove_sub_task(task)
    @sub_tasks.delete(task)
  end

  def get_time_required
    time = 0.0
    @sub_tasks.each { |task| time += task.get_time_required }
    time
  end
end

class AddDryIngredients < Task

  def initialize
    super('Add dry Ingredients')
    puts 'adding dry ingredients'
  end

  def get_time_required
    1.0
  end
end

class Mix < Task

  def initialize
    super('Mix that batter up!')
    puts 'mix batter'
  end

  def get_time_required
    3.0
  end
end

class MakeBatterTask < CompositeTask

  def initialize
    super('Make Batter')
    puts 'Begin - Make Batter'
    add_sub_task( AddDryIngredients.new )
    add_sub_task( Mix.new )
    puts 'End - Make Batter'
  end
end

mbt = MakeBatterTask.new()

File: src/composite/Composite.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace PatternCatalog
{
    public class Composite
    {
        public void Main()
        {
            var composite = new CompositeTask("Make Batter");

            composite.AddItem(new Task("Add Dry Ingredients",TimeSpan.FromMinutes(1)));
            composite.AddItem(new Task("Mix Ingredients", TimeSpan.FromMinutes(3)));

            Console.WriteLine(composite.GetTimeRequired());
        }

        public interface IComponent
        {
            TimeSpan GetTimeRequired();
        }

        public interface IComposite
        {
            void AddItem(IComponent component);
            void RemoveItem(IComponent component);
        }

        // Task is also referred to as a Component
        public class Task : IComponent
        {
            public string _name;
            public TimeSpan _timeRequired;

            public Task(string name, TimeSpan timeRequired)
            {
                _name = name;
                _timeRequired = timeRequired;
            }

            public TimeSpan GetTimeRequired()
            {
                return _timeRequired;
            }
        }

        public class CompositeTask : IComponent, IComposite
        {
            public List<IComponent> SubTasks { get; set; }

            public CompositeTask(string name)
            {
                SubTasks = new List<IComponent>();
            }

            public TimeSpan GetTimeRequired()
            {
                return SubTasks.Aggregate(
                    TimeSpan.FromSeconds(0),
                    (total, comp) => total += comp.GetTimeRequired()
                    );
            }

            public void AddItem(IComponent component)
            {
                SubTasks.Add(component);
            }

            public void RemoveItem(IComponent component)
            {
                SubTasks.Remove(component);
            }
        }
    }
}
